File: backend/old/src/services/expressionService.ts
/**
 * Expression表操作服务
 * 提供Expression表的所有数据库操作功能
 */

import * as path from 'path';
import { databaseManager } from '../core/database';
import { pathCacheManager } from '../core/pathCacheManager';
import { logger } from '../core/logger';
import { 
  ExpressionRecord, 
  ExpressionInsertData, 
  ExpressionUpdateData,
  ExpressionPaginationParams,
  ExpressionPaginationResult,
  ExpressionStats,
  ExpressionType
} from '../types/expression';
import { 
  ValidationError, 
  NotFoundError, 
  InternalServerError 
} from '../middleware/errorHandler';

export class ExpressionService {
  private static readonly TABLE_NAME = 'expression';
  private static readonly CONNECTION_NAME = 'maimai_db';

  /**
   * 初始化数据库连接
   */
  private static async initConnection(): Promise<void> {
    const pathCache = await pathCacheManager.getAllPaths();
    const mainRoot = pathCache.mainRoot;

    if (!mainRoot) {
      throw new NotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存');
    }

    const dbPath = path.join(mainRoot, 'data', 'MaiBot.db');
    
    // 检查连接是否已存在
    if (!databaseManager.hasConnection(this.CONNECTION_NAME)) {
      if (!databaseManager.isManagerInitialized()) {
        await databaseManager.initialize();
      }
      
      await databaseManager.addDatabase(this.CONNECTION_NAME, {
        path: dbPath,
        readonly: false,
        timeout: 5000,
        fileMustExist: true
      });
      
      logger.info('麦麦数据库连接已建立');
    }
  }

  /**
   * 验证Expression数据
   */
  private static validateExpressionData(data: ExpressionInsertData | ExpressionUpdateData): void {
    const errors: string[] = [];

    // 验证situation
    if ('situation' in data && data.situation !== undefined) {
      if (!data.situation || typeof data.situation !== 'string') {
        errors.push('situation必须是非空字符串');
      } else if (data.situation.length > 500) {
        errors.push('situation长度不能超过500字符');
      }
    }

    // 验证style
    if ('style' in data && data.style !== undefined) {
      if (!data.style || typeof data.style !== 'string') {
        errors.push('style必须是非空字符串');
      } else if (data.style.length > 1000) {
        errors.push('style长度不能超过1000字符');
      }
    }

    // 验证count
    if ('count' in data && data.count !== undefined) {
      if (typeof data.count !== 'number' || data.count < 0) {
        errors.push('count必须是非负数字');
      }
    }

    // 验证chat_id
    if ('chat_id' in data && data.chat_id !== undefined) {
      if (!data.chat_id || typeof data.chat_id !== 'string') {
        errors.push('chat_id必须是非空字符串');
      } else if (data.chat_id.length > 100) {
        errors.push('chat_id长度不能超过100字符');
      }
    }

    // 验证type
    if ('type' in data && data.type !== undefined) {
      if (!data.type || typeof data.type !== 'string') {
        errors.push('type必须是非空字符串');
      } else if (!Object.values(ExpressionType).includes(data.type as ExpressionType)) {
        errors.push('type必须是style或grammar');
      }
    }

    // 验证时间字段
    if ('last_active_time' in data && data.last_active_time !== undefined) {
      if (typeof data.last_active_time !== 'number' || data.last_active_time < 0) {
        errors.push('last_active_time必须是非负数字');
      }
    }

    if ('create_date' in data && data.create_date !== undefined) {
      if (typeof data.create_date !== 'number' || data.create_date < 0) {
        errors.push('create_date必须是非负数字');
      }
    }

    // 如果是插入数据，验证必需字段
    if (!('id' in data)) {
      const insertData = data as ExpressionInsertData;
      if (!insertData.situation) errors.push('situation是必需字段');
      if (!insertData.style) errors.push('style是必需字段');
      if (!insertData.chat_id) errors.push('chat_id是必需字段');
      if (!insertData.type) errors.push('type是必需字段');
    }

    if (errors.length > 0) {
      throw new ValidationError(errors.join('; '));
    }
  }

  /**
   * 分页查询expression
   */
  public static async getExpressions(params: ExpressionPaginationParams): Promise<ExpressionPaginationResult> {
    try {
      // 验证分页参数
      if (params.page < 1) {
        throw new ValidationError('页码必须大于0');
      }
      if (params.pageSize < 1 || params.pageSize > 100) {
        throw new ValidationError('每页记录数必须在1-100之间');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 构建查询条件
      const conditions: any[] = [];
      
      if (params.filter) {
        const filter = params.filter;
        
        if (filter.situation) {
          conditions.push({ field: 'situation', operator: 'LIKE', value: `%${filter.situation}%` });
        }
        if (filter.style) {
          conditions.push({ field: 'style', operator: 'LIKE', value: `%${filter.style}%` });
        }
        if (filter.chat_id) {
          conditions.push({ field: 'chat_id', operator: '=', value: filter.chat_id });
        }
        if (filter.type) {
          conditions.push({ field: 'type', operator: '=', value: filter.type });
        }
        if (filter.minCount !== undefined) {
          conditions.push({ field: 'count', operator: '>=', value: filter.minCount });
        }
        if (filter.maxCount !== undefined) {
          conditions.push({ field: 'count', operator: '<=', value: filter.maxCount });
        }
        if (filter.startDate !== undefined) {
          conditions.push({ field: 'create_date', operator: '>=', value: filter.startDate });
        }
        if (filter.endDate !== undefined) {
          conditions.push({ field: 'create_date', operator: '<=', value: filter.endDate });
        }
      }

      // 获取总记录数
      const totalCount = await operator.count(this.TABLE_NAME, conditions.length > 0 ? { where: conditions } : {});

      // 计算分页信息
      const totalPages = Math.ceil(totalCount / params.pageSize);
      const offset = (params.page - 1) * params.pageSize;

      // 查询当前页数据
      const queryOptions: any = {
        limit: params.pageSize,
        offset: offset,
        orderBy: [{ 
          field: params.orderBy || 'id', 
          order: params.orderDir || 'ASC' 
        }]
      };

      if (conditions.length > 0) {
        queryOptions.where = conditions;
      }

      const items = await operator.findMany<ExpressionRecord>(this.TABLE_NAME, queryOptions);

      const result: ExpressionPaginationResult = {
        items,
        total: totalCount,
        totalPages,
        currentPage: params.page,
        pageSize: params.pageSize,
        hasNext: params.page < totalPages,
        hasPrev: params.page > 1
      };

      return result;

    } catch (error) {
      logger.error('分页查询expression失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询expression失败');
    }
  }

  /**
   * 根据ID查询单个expression
   */
  public static async getExpressionById(id: number): Promise<ExpressionRecord | null> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findOne<ExpressionRecord>(this.TABLE_NAME, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      return result;

    } catch (error) {
      logger.error('根据ID查询expression失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询expression失败');
    }
  }

  /**
   * 插入expression
   */
  public static async insertExpression(data: ExpressionInsertData): Promise<number> {
    try {
      this.validateExpressionData(data);

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 设置默认值（只对用户未提供的字段设置默认值）
      const currentTime = Date.now() / 1000;
      const defaults = {
        count: 0,
        last_active_time: currentTime,
        create_date: currentTime,
      };
      
      // 合并数据，用户提供的数据优先
      const insertData = { ...defaults, ...data };

      const result = await operator.insertOne(this.TABLE_NAME, insertData);

      if (!result.lastInsertId) {
        throw new InternalServerError('插入expression失败，未获得插入ID');
      }

      logger.info(`expression插入成功，ID: ${result.lastInsertId}`);
      return result.lastInsertId;

    } catch (error) {
      logger.error('插入expression失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('插入expression失败');
    }
  }

  /**
   * 更新expression
   */
  public static async updateExpression(data: ExpressionUpdateData): Promise<boolean> {
    try {
      if (!data.id || data.id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      this.validateExpressionData(data);

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 检查记录是否存在
      const existingRecord = await this.getExpressionById(data.id);
      if (!existingRecord) {
        throw new NotFoundError('要更新的expression记录不存在');
      }

      // 提取更新数据（排除id）
      const { id, ...updateData } = data;

      const result = await operator.update(this.TABLE_NAME, updateData, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      if (result.affectedRows === 0) {
        throw new InternalServerError('更新expression失败，未影响任何记录');
      }

      logger.info(`expression更新成功，ID: ${id}`);
      return true;

    } catch (error) {
      logger.error('更新expression失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('更新expression失败');
    }
  }

  /**
   * 删除expression
   */
  public static async deleteExpression(id: number): Promise<boolean> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 检查记录是否存在
      const existingRecord = await this.getExpressionById(id);
      if (!existingRecord) {
        throw new NotFoundError('要删除的expression记录不存在');
      }

      const result = await operator.delete(this.TABLE_NAME, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      if (result.affectedRows === 0) {
        throw new InternalServerError('删除expression失败，未影响任何记录');
      }

      logger.info(`expression删除成功，ID: ${id}`);
      return true;

    } catch (error) {
      logger.error('删除expression失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('删除expression失败');
    }
  }

  /**
   * 增加统计次数
   */
  public static async incrementCount(id: number): Promise<void> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 使用原生SQL更新统计次数和最后活跃时间
      await operator.executeStatement(
        'UPDATE expression SET count = count + 1, last_active_time = ? WHERE id = ?',
        [Date.now() / 1000, id]
      );

    } catch (error) {
      logger.error('增加统计次数失败:', error);
      // 这个操作失败不应该影响主要业务流程，只记录错误
    }
  }

  /**
   * 根据聊天ID查询expression
   */
  public static async getExpressionsByChatId(chatId: string, limit: number = 10): Promise<ExpressionRecord[]> {
    try {
      if (!chatId) {
        throw new ValidationError('聊天ID不能为空');
      }

      if (limit < 1 || limit > 100) {
        throw new ValidationError('限制数量必须在1-100之间');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findMany<ExpressionRecord>(this.TABLE_NAME, {
        where: [{ field: 'chat_id', operator: '=', value: chatId }],
        orderBy: [{ field: 'last_active_time', order: 'DESC' }],
        limit
      });

      return result;

    } catch (error) {
      logger.error('根据聊天ID查询expression失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询expression失败');
    }
  }

  /**
   * 根据类型查询expression
   */
  public static async getExpressionsByType(type: string, limit: number = 10): Promise<ExpressionRecord[]> {
    try {
      if (!type) {
        throw new ValidationError('类型不能为空');
      }

      if (!Object.values(ExpressionType).includes(type as ExpressionType)) {
        throw new ValidationError('类型必须是style或grammar');
      }

      if (limit < 1 || limit > 100) {
        throw new ValidationError('限制数量必须在1-100之间');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findMany<ExpressionRecord>(this.TABLE_NAME, {
        where: [{ field: 'type', operator: '=', value: type }],
        orderBy: [{ field: 'count', order: 'DESC' }],
        limit
      });

      return result;

    } catch (error) {
      logger.error('根据类型查询expression失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询expression失败');
    }
  }

  /**
   * 获取expression统计信息
   */
  public static async getExpressionStats(): Promise<ExpressionStats> {
    try {
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 获取总数
      const total = await operator.count(this.TABLE_NAME, {});

      // 获取按类型统计
      const typeStats = await operator.executeQuery<{ type: string; count: number }>(
        'SELECT type, COUNT(*) as count FROM expression GROUP BY type'
      );

      // 获取按聊天ID统计（取前10个最活跃的）
      const chatStats = await operator.executeQuery<{ chat_id: string; count: number }>(
        'SELECT chat_id, COUNT(*) as count FROM expression GROUP BY chat_id ORDER BY count DESC LIMIT 10'
      );

      // 获取平均统计次数和总统计次数
      const countStats = await operator.executeQuery<{ avgCount: number; totalCount: number }>(
        'SELECT AVG(count) as avgCount, SUM(count) as totalCount FROM expression'
      );

      // 获取最近活跃的expression数量（24小时内）
      const dayAgo = (Date.now() / 1000) - (24 * 60 * 60);
      const recentActive = await operator.count(this.TABLE_NAME, {
        where: [{ field: 'last_active_time', operator: '>=', value: dayAgo }]
      });

      // 处理统计结果
      const byType: Record<string, number> = {};
      typeStats.forEach((row) => {
        byType[row.type] = row.count;
      });

      const byChatId: Record<string, number> = {};
      chatStats.forEach((row) => {
        byChatId[row.chat_id] = row.count;
      });

      const avgCount = countStats[0]?.avgCount || 0;
      const totalCount = countStats[0]?.totalCount || 0;

      const result: ExpressionStats = {
        total,
        byType,
        byChatId,
        avgCount: Math.round(avgCount * 100) / 100, // 保留2位小数
        totalCount,
        recentActive
      };

      return result;

    } catch (error) {
      logger.error('获取expression统计信息失败:', error);
      throw new InternalServerError('获取expression统计信息失败');
    }
  }

  /**
   * 搜索expression（按情境和风格）
   */
  public static async searchExpressions(
    keyword: string, 
    limit: number = 20
  ): Promise<ExpressionRecord[]> {
    try {
      if (!keyword || keyword.trim().length === 0) {
        throw new ValidationError('搜索关键字不能为空');
      }

      if (limit < 1 || limit > 100) {
        throw new ValidationError('限制数量必须在1-100之间');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 使用原生SQL查询来实现OR逻辑
      const result = await operator.executeQuery<ExpressionRecord>(
        `SELECT * FROM ${this.TABLE_NAME} 
         WHERE situation LIKE ? OR style LIKE ? 
         ORDER BY count DESC, last_active_time DESC 
         LIMIT ?`,
        [`%${keyword}%`, `%${keyword}%`, limit]
      );

      return result;

    } catch (error) {
      logger.error('搜索expression失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('搜索expression失败');
    }
  }
}

File: backend/src/services/expression_service.py
"""
Expression表业务逻辑服务
提供Expression表的所有业务操作功能
"""

import time
from typing import Optional, List

from core.database_manager import database_manager
from core.database_operator import DatabaseOperator
from core.logger import logger
from models.database import OrderDirection, QueryParams
from models.expression import (
    ExpressionRecord, ExpressionInsertData, ExpressionUpdateData,
    ExpressionPaginationParams, ExpressionPaginationResult,
    ExpressionStats
)
from utils.database_validator import DatabaseValidator


class ExpressionService:
    """Expression表服务类"""
    
    TABLE_NAME = "expression"
    DATABASE_NAME = "maibot"  # 修正为正确的数据库连接名
    
    @classmethod
    async def get_expressions(cls, params: ExpressionPaginationParams) -> ExpressionPaginationResult:
        """
        分页查询expression列表
        
        Args:
            params: 分页查询参数
            
        Returns:
            分页查询结果
        """
        try:
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 构建WHERE条件
            where_conditions = {}
            if params.filter:
                if params.filter.situation:
                    where_conditions["situation LIKE"] = f"%{params.filter.situation}%"
                if params.filter.style:
                    where_conditions["style LIKE"] = f"%{params.filter.style}%"
                if params.filter.chat_id:
                    where_conditions["chat_id"] = params.filter.chat_id
                if params.filter.type:
                    where_conditions["type"] = params.filter.type
                if params.filter.minCount is not None:
                    where_conditions["count >="] = params.filter.minCount
                if params.filter.maxCount is not None:
                    where_conditions["count <="] = params.filter.maxCount
                if params.filter.startDate is not None:
                    where_conditions["create_date >="] = params.filter.startDate
                if params.filter.endDate is not None:
                    where_conditions["create_date <="] = params.filter.endDate
            
            # 查询数据
            result = operator.find_with_pagination(
                table_name=cls.TABLE_NAME,
                where_conditions=where_conditions,
                page=params.page,
                page_size=params.page_size,
                order_by=params.orderBy or "id",
                order_dir=params.orderDir or "ASC"
            )
            
            # 转换为Expression记录
            items = []
            for row in result.items:
                items.append(ExpressionRecord(**row))
            
            return ExpressionPaginationResult(
                items=items,
                total=result.total,
                page=result.current_page,
                size=result.page_size,
                totalPages=result.total_pages
            )
            
        except Exception as error:
            logger.error(f"查询expression列表失败: {error}")
            raise
    
    @classmethod
    async def get_expression_by_id(cls, expression_id: int) -> Optional[ExpressionRecord]:
        """
        根据ID获取单个expression
        
        Args:
            expression_id: expression ID
            
        Returns:
            expression记录，未找到时返回None
        """
        try:
            if expression_id <= 0:
                raise ValueError("expression ID必须大于0")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 查询单条记录
            result = operator.find_one(
                table_name=cls.TABLE_NAME,
                where_conditions={"id": expression_id}
            )
            
            if result:
                return ExpressionRecord(**result)
            
            return None
            
        except Exception as error:
            logger.error(f"根据ID查询expression失败: {error}")
            raise
    
    @classmethod
    async def insert_expression(cls, data: ExpressionInsertData) -> int:
        """
        插入新的expression记录
        
        Args:
            data: 插入数据
            
        Returns:
            新插入记录的ID
        """
        try:
            # 验证数据
            DatabaseValidator.validate_not_empty(data.situation, "situation")
            DatabaseValidator.validate_not_empty(data.style, "style")
            DatabaseValidator.validate_not_empty(data.chat_id, "chat_id")
            DatabaseValidator.validate_not_empty(data.type, "type")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 准备插入数据
            current_time = time.time()
            insert_data = {
                "situation": data.situation,
                "style": data.style,
                "count": data.count or 0.0,
                "last_active_time": data.last_active_time or current_time,
                "chat_id": data.chat_id,
                "type": data.type,
                "create_date": data.create_date or current_time
            }
            
            # 执行插入
            result = operator.insert(
                table_name=cls.TABLE_NAME,
                data=insert_data
            )
            
            if result.affected_rows > 0:
                logger.info(f"expression插入成功，ID: {result.last_insert_id}")
                return result.last_insert_id
            else:
                raise RuntimeError("插入expression失败")
                
        except Exception as error:
            logger.error(f"插入expression失败: {error}")
            raise
    
    @classmethod
    async def update_expression(cls, data: ExpressionUpdateData) -> bool:
        """
        更新expression记录
        
        Args:
            data: 更新数据
            
        Returns:
            是否更新成功
        """
        try:
            if data.id <= 0:
                raise ValueError("expression ID必须大于0")
            
            # 检查记录是否存在
            existing = await cls.get_expression_by_id(data.id)
            if not existing:
                raise RuntimeError(f"未找到ID为 {data.id} 的expression记录")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 准备更新数据（只包含有值的字段）
            update_data = {}
            if data.situation is not None:
                DatabaseValidator.validate_not_empty(data.situation, "situation")
                update_data["situation"] = data.situation
            if data.style is not None:
                DatabaseValidator.validate_not_empty(data.style, "style")
                update_data["style"] = data.style
            if data.count is not None:
                update_data["count"] = data.count
            if data.last_active_time is not None:
                update_data["last_active_time"] = data.last_active_time
            if data.chat_id is not None:
                DatabaseValidator.validate_not_empty(data.chat_id, "chat_id")
                update_data["chat_id"] = data.chat_id
            if data.type is not None:
                DatabaseValidator.validate_not_empty(data.type, "type")
                update_data["type"] = data.type
            if data.create_date is not None:
                update_data["create_date"] = data.create_date
                
            if not update_data:
                logger.warning("没有需要更新的数据")
                return False
                
            # 执行更新
            result = operator.update(
                table_name=cls.TABLE_NAME,
                data=update_data,
                where_conditions={"id": data.id}
            )
            
            if result.affected_rows > 0:
                logger.info(f"expression更新成功，ID: {data.id}")
                return True
            else:
                logger.warning(f"未找到要更新的expression，ID: {data.id}")
                return False
                
        except Exception as error:
            logger.error(f"更新expression失败: {error}")
            raise
    
    @classmethod
    async def delete_expression(cls, expression_id: int) -> bool:
        """
        删除expression记录
        
        Args:
            expression_id: expression ID
            
        Returns:
            是否删除成功
        """
        try:
            if expression_id <= 0:
                raise ValueError("expression ID必须大于0")
            
            # 检查记录是否存在
            existing = await cls.get_expression_by_id(expression_id)
            if not existing:
                raise RuntimeError(f"未找到ID为 {expression_id} 的expression记录")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 执行删除
            result = operator.delete(
                table_name=cls.TABLE_NAME,
                where_conditions={"id": expression_id}
            )
            
            if result.affected_rows > 0:
                logger.info(f"expression删除成功，ID: {expression_id}")
                return True
            else:
                logger.warning(f"未找到要删除的expression，ID: {expression_id}")
                return False
                
        except Exception as error:
            logger.error(f"删除expression失败: {error}")
            raise
    
    @classmethod
    async def get_expressions_by_chat_id(cls, chat_id: str, limit: int = 10) -> List[ExpressionRecord]:
        """
        根据聊天ID查询expression
        
        Args:
            chat_id: 聊天ID
            limit: 限制结果数量
            
        Returns:
            expression记录列表
        """
        try:
            if not chat_id or not chat_id.strip():
                raise ValueError("聊天ID不能为空")
            
            if limit <= 0 or limit > 100:
                raise ValueError("限制数量必须在1-100之间")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 查询数据
            results = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    where={"chat_id": chat_id.strip()},
                    order_by="last_active_time",
                    order_dir=OrderDirection.DESC,
                    limit=limit
                )
            )
            
            # 转换为Expression记录
            items = []
            for row in results:
                items.append(ExpressionRecord(**row))
            
            return items
            
        except Exception as error:
            logger.error(f"根据聊天ID查询expression失败: {error}")
            raise
    
    @classmethod
    async def get_expressions_by_type(cls, expr_type: str, limit: int = 10) -> List[ExpressionRecord]:
        """
        根据类型查询expression
        
        Args:
            expr_type: 类型
            limit: 限制结果数量
            
        Returns:
            expression记录列表
        """
        try:
            if not expr_type or not expr_type.strip():
                raise ValueError("类型不能为空")
            
            if limit <= 0 or limit > 100:
                raise ValueError("限制数量必须在1-100之间")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 查询数据
            results = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    where={"type": expr_type.strip()},
                    order_by="count",
                    order_dir=OrderDirection.DESC,
                    limit=limit
                )
            )
            
            # 转换为Expression记录
            items = []
            for row in results:
                items.append(ExpressionRecord(**row))
            
            return items
            
        except Exception as error:
            logger.error(f"根据类型查询expression失败: {error}")
            raise
    
    @classmethod
    async def search_expressions(cls, keyword: str, limit: int = 20) -> List[ExpressionRecord]:
        """
        搜索expression
        
        Args:
            keyword: 搜索关键字
            limit: 限制结果数量
            
        Returns:
            expression记录列表
        """
        try:
            if not keyword or not keyword.strip():
                raise ValueError("搜索关键字不能为空")
            
            if limit <= 0 or limit > 100:
                raise ValueError("限制数量必须在1-100之间")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 构建搜索条件
            keyword = keyword.strip()
            where_conditions = {
                "OR": [
                    {"situation LIKE": f"%{keyword}%"},
                    {"style LIKE": f"%{keyword}%"}
                ]
            }
            
            # 查询数据
            results = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    where=where_conditions,
                    order_by="count",
                    order_dir=OrderDirection.DESC,
                    limit=limit
                )
            )
            
            # 转换为Expression记录
            items = []
            for row in results:
                items.append(ExpressionRecord(**row))
            
            return items
            
        except Exception as error:
            logger.error(f"搜索expression失败: {error}")
            raise
    
    @classmethod
    async def get_expression_stats(cls) -> ExpressionStats:
        """
        获取expression统计信息
        
        Returns:
            统计信息
        """
        try:
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 获取总记录数
            total_result = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    select=["COUNT(*) as total"],
                    where={}
                )
            )
            total = total_result[0]["total"] if total_result else 0
            
            # 按类型统计
            type_result = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    select=["type", "COUNT(*) as count"],
                    where={}
                )
            )
            by_type = {row["type"]: row["count"] for row in type_result}
            
            # 按聊天ID统计（取前10个）
            chat_result = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    select=["chat_id", "COUNT(*) as count"],
                    where={},
                    limit=10,
                    order_by="count",
                    order_dir=OrderDirection.DESC
                )
            )
            by_chat_id = {row["chat_id"]: row["count"] for row in chat_result}
            
            # 统计次数相关
            count_result = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    select=["AVG(count) as avg_count", "SUM(count) as total_count"],
                    where={}
                )
            )
            avg_count = float(count_result[0]["avg_count"] or 0) if count_result else 0.0
            total_count = float(count_result[0]["total_count"] or 0) if count_result else 0.0
            
            # 最近活跃的expression数量（24小时内）
            recent_time = time.time() - 24 * 60 * 60  # 24小时前
            recent_result = operator.find_many(
                table_name=cls.TABLE_NAME,
                params=QueryParams(
                    select=["COUNT(*) as recent_count"],
                    where={"last_active_time >=": recent_time}
                )
            )
            recent_active = recent_result[0]["recent_count"] if recent_result else 0
            
            return ExpressionStats(
                total=total,
                byType=by_type,
                byChatId=by_chat_id,
                avgCount=avg_count,
                totalCount=total_count,
                recentActive=recent_active
            )
            
        except Exception as error:
            logger.error(f"获取expression统计信息失败: {error}")
            raise
    
    @classmethod
    async def increment_count(cls, expression_id: int) -> bool:
        """
        增加expression统计次数
        
        Args:
            expression_id: expression ID
            
        Returns:
            是否更新成功
        """
        try:
            if expression_id <= 0:
                raise ValueError("expression ID必须大于0")
            
            # 获取数据库连接
            db_connection = database_manager.get_connection(cls.DATABASE_NAME)
            if not db_connection:
                raise RuntimeError("数据库连接未初始化")
            
            operator = DatabaseOperator(db_connection)
            
            # 更新统计次数和最后活跃时间
            current_time = time.time()
            update_data = {
                "count": "count + 1",  # 使用SQL表达式
                "last_active_time": current_time
            }
            
            result = operator.update(
                table_name=cls.TABLE_NAME,
                data=update_data,
                where_conditions={"id": expression_id}
            )
            
            if result.affected_rows > 0:
                logger.info(f"expression统计次数更新成功，ID: {expression_id}")
                return True
            else:
                logger.warning(f"未找到要更新的expression，ID: {expression_id}")
                return False
                
        except Exception as error:
            logger.error(f"更新expression统计次数失败: {error}")
            raise
